### PR TITLE
Feat/gunnraj d/#1 database and models

### PR DESCRIPTION
**Pull request by Gunnraj-D**
# Task
Establishes SQLite database and models, with methods to interact with the database in predefined ways.

# Root Cause
We currenlty do not have a database.

# Solution 
2 python classes that represent a project and transcription. These classes have methods that allow you to interact with the database.

# How Has This Been Tested? 
Rohith made test cases using unittests. First enter bash terminal by typing "bash" into your terminal in vscode. Then navigate to the server directory. Then run the command "python -m unittest". This will execute the tests.

# Checklist: 
- [ ] I have linked issue to this Pull Request 
- [✔] I have performed a self-review of my code 
- [ ✔] I have commented my code, particularly in hard-to-understand areas 
- [ ] I have made corresponding changes to the documentation 
- [ ✔] I have added tests that prove my fix is effective or that my feature works 
- [✔ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: server/app/database_models.py [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+                cur.execute(f"""
+                    INSERT INTO {PROJECT_TABLE_NAME} (name, description) VALUES (?, ?)
+                """, (project_name, description))
+
+                project_id =  cur.lastrowid
+
+            return project_id
+
+        except sqlite3.Error as e:
+            print(f"There was an error inserting into the table: {e}")
```
Comment (YueYaoYan) [inline server/app/database_models.py]: It might be better to just throw an exception here

File: server/app/database_models.py [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+
+                cur.execute(f"""
+                    DELETE FROM {PROJECT_TABLE_NAME}
+                    WHERE project_id = ? 
+                """, (project_id,))
+
+            return True
+
+        except sqlite3.Error as e:
+            print(f"There was an error deleting the project: {e}")
```
Comment (YueYaoYan) [inline server/app/database_models.py]: Same here

File: server/app/database_models.py [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+
+            return None
+
+        except sqlite3.Error as e:
+            print(f"There was an error getting the projects: {e}")
+            return None
+
+
+
+class Transcription:
```
Comment (YueYaoYan) [inline server/app/database_models.py:140]: What you have now is great because we only have two table refactoring here is not really neccessary but I'll put the suggestions here and leave the decision up to you.

Since these two classes are pretty similar, like `_create_table`, `insert`, `delete`, `drop_table`. I think it might be better to create an abstract base class for these common database operation and let these two class inherent the parent class. For example, like `_create_table` can be like:
```def _create_table(self, table_name: str, schema: str) -> bool:
        """Create a table with the given schema"""
        try:
            self._execute(f"CREATE TABLE IF NOT EXISTS {table_name} ({schema})")
            return True
        except sqlite3.Error as e:
             ...
 ```

File: server/app/database_models.py
Comment (guiltyeaten2) [inline server/app/database_models.py:1]: Well written database model for project and transcription, code are clear and very readable and the code takes security into considerations.
 
Good use of the global attribute, ensure the readability of the code and simplify usage.
Good documentation overall.
Good error handling on each operation function. 

Good use of "?" as place holder and passing the argument as a tuple element in cursor execution. This ensure the security of the database, prevent potential SQL-injection attack as SQLite would escape the uncommon input. Also this method ensure the consistency of the code especially in the case of passing multiple argument into schema.

Overall, the code are consistence, correct and easy to understand.

File: server/tests/__init__.py
Comment (guiltyeaten2) [inline server/tests/__init__.py:1]: Not sure why this empty file exist. May need to removed this file if unneeded.

File: server/app/database_models.py [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+        
+
+    def _create_table(self) -> None:
+        """
+        Private method. Creates the project table if it does not yet exist, otherwise does nothing.
+        """
+        try:
+            with sqlite3.connect(self.db_name) as conn:
+                cur = conn.cursor()
+                cur.execute("PRAGMA foreign_keys = ON;")
```
Comment (guiltyeaten2) [inline server/app/database_models.py]: Good use of " cur.execute("PRAGMA foreign_keys = ON;")" ensure the foreign constrain always on when doing insertion and deletion. as this feature were set to off by default in SQLite.

File: server/app/database_models.py [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+        self.db_name = database_name
+        self._create_table()
+        
+
+    def _create_table(self) -> None:
+        """
+        Private method. Creates the project table if it does not yet exist, otherwise does nothing.
+        """
+        try:
+            with sqlite3.connect(self.db_name) as conn:
```
Comment (guiltyeaten2) [inline server/app/database_models.py]: Good handling with the database connection by using "with sqlite3.connect(self.db_name) as conn". This ensure the software only connect to database when do operations and close once finished operation.

File: server/tests/test_databasemodels.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+from app.database_models import Project, Transcription  # Make sure these are correctly imported
+
+DB_NAME = "qualAI_test.db"
+PROJECT_TABLE_NAME = "project"
+
+class Test_Project(unittest.TestCase):
+    def setUp(self):
+        """Set up for each test - ensures a clean database."""
+        try:
+            remove(DB_NAME)
```
Comment (guiltyeaten2) [inline server/tests/test_databasemodels.py:15]: Deleting the entire SQL file does the job, but it's more ideal to create the test SQL db if the db is not existed, and use SQLite3 command to clean the existed test SQL db. Either way, this approach would work and should not causing issue as long as the DB_NAME  above is different from the main SQL db filename

File: server/tests/test_databasemodels.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+        except FileNotFoundError:
+            pass
+        self.project_manager = Project(DB_NAME)
+
+    def test_insert_project(self):
```
Comment (guiltyeaten2) [inline server/tests/test_databasemodels.py]: maybe test the invalid insert as well? apparently this only test the success case. Maybe add another insert that's passing null value for both and see what happen. (this might be edge case but i guess it still would be good to test it)

File: server/tests/test_databasemodels.py [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+
+    def test_get_project_by_id_not_found(self):
+        project_name = "unique_project_3"
+        inserted_id = self.project_manager.insert(project_name, "Yet another description")
+        self.assertIsNotNone(inserted_id)
+        non_existent_id = inserted_id + 100
+        retrieved_project = self.project_manager.get_project_by_id(non_existent_id)
+        self.assertIsNone(retrieved_project)
+
+    def test_delete_project(self):
```
Comment (guiltyeaten2) [inline server/tests/test_databasemodels.py:49]: similar here, maybe test the case to delete an none existed file as well

File: server/tests/test_databasemodels.py [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+        self.assertEqual(len(all_projects), 2)
+        names = {proj[1] for proj in all_projects}
+        self.assertIn("unique_project_5", names)
+        self.assertIn("unique_project_6", names)
+
+class Test_Transcription(unittest.TestCase):
+    def setUp(self):
+        """Set up for each test ."""
+        try:
+            remove(DB_NAME)
```
Comment (guiltyeaten2) [inline server/tests/test_databasemodels.py:77]: same issue as above when testing project

File: server/tests/test_databasemodels.py [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+        self.assertEqual(transcription[2], content) # Check content (index 2)
+
+    """ Test to check a transcription that doesnt exist via a non valid id"""
+    def test_get_transcription_id_not_found(self):
+        non_existent_id = 999
+        retrieved_transcription = self.transcription_manager.get_transcription_by_id(non_existent_id)
+        self.assertIsNone(retrieved_transcription)
+
+    """ test to check valid deletion of transcription in table"""
+    def test_delete_transcription(self):
```
Comment (guiltyeaten2) [inline server/tests/test_databasemodels.py:134]: same as above, maybe try to cover more test case

File: server/tests/test_databasemodels.py [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+            pass
+        self.project_manager = Project(DB_NAME)
+        self.transcription_manager = Transcription(DB_NAME)
+
+        # Ensure there is atleast one project for transcription tests
+        self.test_project_id = self.project_manager.insert("base_project_for_transcriptions", "Base project")
+        self.assertIsNotNone(self.test_project_id, "Failed to create a base project for transcription tests")
+
+    """Test to check if you can insert a transcription to the table"""
+    def test_insert_transcription(self):
```
Comment (guiltyeaten2) [inline server/tests/test_databasemodels.py:92]: same as above, maybe try to cover more test case

File: server/tests/test_databasemodels.py
Comment (guiltyeaten2) [inline server/tests/test_databasemodels.py:1]: Well written unittest for database model. while there are some minor issue when maintaining the testing database and some of the potential output were not be covered in the test case, this file can still test the database model mostly and ensure the functionality of the model in most scenario

**Review by guiltyeaten2 — APPROVED** [review]
Done code review for database_model.py and the corresponding test file. Overall good approach were taken and there's only a few minor issue in the test file, but those issue won't affect the major functionality of the software so it's okay to be merged.

**Review by YueYaoYan — APPROVED** [review]

**Review by YueYaoYan — APPROVED** [review]
Looks good now! If you could also add a README.md for how to run, call your db with some example it would be great! (but this can be done split into another issue for later too)
